feat: perform selected strategy for isbsvc fields to support PPND

dpadhiar opened the pull request:
<!-- markdownlint-disable MD041 -->

<!-- Does this PR fix an issue -->

Fixes #248 

### Modifications

Integrates the USDE logic to derive an upgrade strategy (that we currently have for PipelineRollouts) into the ISBServiceRollout reconciliation process.

### Verification

Ran `make test` to verify unit tests are passing.

Ran `make test-e2e` locally and on CI to see that updating with and without excluded fields works properly.

juliev0 commented on tests/e2e/functional_test.go:
```diff
@@ -524,6 +541,56 @@ var _ = Describe("Functional e2e", Serial, func() {
+				verifyPipelineStatusConsistently(Namespace, pipelineName, func(retrievedPipelineSpec numaflowv1.PipelineSpec, retrievedPipelineStatus numaflowv1.PipelineStatus) bool {
+					return retrievedPipelineStatus.Phase != numaflowv1.PipelinePhasePaused
+				})
+			}
```
looks like `if dataLossPrevention == "true"` is in here twice

dpadhiar replied:
Oops, fixing that now.

juliev0 commented on tests/e2e/functional_test.go:
```diff
@@ -524,6 +541,56 @@ var _ = Describe("Functional e2e", Serial, func() {
+		if dataLossPrevention == "true" {
+
+			document("Verify that in-progress-strategy gets set to NoOp")
+			verifyInProgressStrategyISBService(Namespace, isbServiceRolloutName, apiv1.UpgradeStrategyNoOp)
+			verifyInProgressStrategy(pipelineRolloutName, apiv1.UpgradeStrategyNoOp)
```
hmm, just thinking that these may have no effect, as they may have already been "NoOp"

dpadhiar replied:
This is true, is there no chance it could be set to PauseAndDrain incorrectly? If not, can remove these checks.

juliev0 replied:
or maybe you can group it with the same "consistently" code I mention below?

juliev0 replied:
I just realized the two statements above `verifyInProgressStrategyISBService` and `verifyInProgressStrategy` are `Eventually` statements themselves - we want to avoid those being within a `Consistently`.

juliev0 commented on tests/e2e/functional_test.go:
```diff
@@ -524,6 +541,54 @@ var _ = Describe("Functional e2e", Serial, func() {
+					return false
+				}
+				return true
+			}, testTimeout).Should(BeTrue())
```
We really need to do this "consistently", not "eventually", right? We kind of need to verify that lines 562-575 are happening consistently.

dpadhiar replied:
That sounds right, how long should we check - 30 seconds? 1 minute?

juliev0 replied:
How about 20 sec or something around there?

dpadhiar replied:
So it looks like consistently will immediately fail as soon as the pausing pipelines condition is true unlike eventually which waits till they're both false. That's why it keeps failing quickly.

dpadhiar replied:
Taking a look at the logs.. the Pipeline does pause. That shouldn't be happening in this case, hm.

juliev0 replied:
> So it looks like consistently will immediately fail as soon as the pausing pipelines condition is true unlike eventually which waits till they're both false

Yes, when we check `Consistently`, we do need to check `Eventually` first so we can make sure it got to that state in the first place

juliev0 replied:
Ideally we can first "Eventually" and then "Consistently" check:

- verifyInProgressStrategyISBService(Namespace, isbServiceRolloutName, apiv1.UpgradeStrategyNoOp)
- verifyInProgressStrategy(pipelineRolloutName, apiv1.UpgradeStrategyNoOp)
- PipelineStatus != Pausing or Paused
- PausingConditions false

To avoid repeating a large chunk of code, if we can't think of good functions, we can always create a temporary function to call each time within this function like:
```
verifyNotPausing := func(str string) {
       .....
    }

Consistently(....
   verifyNotPausing
}

Eventually(...
   verifyNotPausing
}
```

juliev0 commented on internal/controller/isbservicerollout_controller.go:
```diff
@@ -278,32 +302,60 @@ func (r *ISBServiceRolloutReconciler) processExistingISBService(ctx context.Cont
 	r.processISBServiceStatus(ctx, existingISBServiceDef, isbServiceRollout)
 
 	// if I need to update or am in the middle of an update of the ISBService, then I need to make sure all the Pipelines are pausing
```
can you update the comment to remove "need to update or"?

juliev0 commented:
nice job

juliev0 commented on tests/e2e/functional_test.go:
```diff
@@ -524,6 +541,50 @@ var _ = Describe("Functional e2e", Serial, func() {
+			return rollout, nil
+		})
+
+		if dataLossPrevention == "true" {
```
nit: I believe the block of code should apply whether or not `dataLossPrevention == true`?

juliev0 commented:
Spoke to @dpadhiar and when he gets time he'll explore why the previous test for updating isbsvc with PPND is sometimes failing before we merge this. (Merging this is not critically important to be done by Q1)

juliev0 commented:
Actually, I went through the artifacts I had when I saw the failure locally - not sure if numaplane is really doing anything wrong - we ultimately need to get the fixes from numaflow and try with that. I'm still a little nervous about merging this in since we weren't seeing this failure before.

juliev0 commented on internal/controller/isbservicerollout_controller.go:
```diff
@@ -277,33 +301,61 @@ func (r *ISBServiceRolloutReconciler) processExistingISBService(ctx context.Cont
+	// if not, should we set one?
+	if !inProgressStrategySet {
+		if userPreferredStrategy == config.PPNDStrategyID {
+			if upgradeStrategyType == apiv1.UpgradeStrategyPPND {
```
nit: I know you copied this from my redundant code in PipelineRolloutController, so my bad, but I believe the check for `userPreferredStrategy` here and below may be unnecessary as it's happening within `ResourceNeedsUpdating()`. We probably don't need to get the value for `userPreferredStrategy` above at all.

juliev0 commented on internal/controller/isbservicerollout_controller_test.go:
```diff
@@ -260,119 +261,133 @@ func Test_reconcile_isbservicerollout_PPND(t *testing.T) {
+			existingStatefulSetDef:     createDefaultISBStatefulSet("2.10.11", true),
+			existingPipeline:           createDefaultPipelineOfPhase(numaflowv1.PipelinePhasePaused, map[string]string{}),
+			expectedRolloutPhase:       apiv1.PhaseDeployed,
+			existingInProgressStrategy: nil,
```
nit: I'm thinking in this case the incoming InProgressStrategy would probably be PPND

juliev0 commented on tests/e2e/functional_test.go:
```diff
@@ -524,6 +547,50 @@ var _ = Describe("Functional e2e", Serial, func() {
+			return *retrievedISBServiceSpec.JetStream.ContainerTemplate.Resources.Limits.Memory() == memLimit
+		})
+
+		verifyISBSvcRolloutReady(isbServiceRolloutName)
+
+		verifyISBSvcReady(Namespace, isbServiceRolloutName, 3)
+
+		verifyInProgressStrategyISBService(Namespace, isbServiceRolloutName, apiv1.UpgradeStrategyNoOp)
+		verifyInProgressStrategy(pipelineRolloutName, apiv1.UpgradeStrategyNoOp)
+		verifyPipelineRunning(Namespace, pipelineName, 3)
+
```
some of these are being checked in the `Consistently` block above I think, and we don't need them ?

dpadhiar replied:
Removed checking the in progress strategy after the Consistently block.

dpadhiar commented on tests/e2e/functional_test.go:
```diff
@@ -524,6 +547,50 @@ var _ = Describe("Functional e2e", Serial, func() {
+			verifyInProgressStrategy(pipelineRolloutName, apiv1.UpgradeStrategyNoOp)
+			verifyPipelineStatusConsistently(Namespace, pipelineName, func(retrievedPipelineSpec numaflowv1.PipelineSpec, retrievedPipelineStatus numaflowv1.PipelineStatus) bool {
+				return retrievedPipelineStatus.Phase != numaflowv1.PipelinePhasePaused
+			})
```
Maybe if we just use Eventually instead of the Consistently version of this function it should work.

juliev0 replied:
yeah, but then it's not truly doing a "consistently" check. The purpose is to verify at every time interval that the whole thing is try. Imagine that the eventually statement inside is blocking for a long period of time.

dpadhiar replied:
Ack, will make this a single check.

juliev0 replied:
What I was saying was that inside the `Consistently` which calls this block, what it should do is not contain any `Eventually` statements at all (or `Consistently` statements). It should simply do a `Get()` on each of the different resources, and check that they are correct every iteration.

juliev0 commented:
FYI, Sidhant says he can get his numaflow changes ready to merge into our testing soon, so then we can see if that intermittent failure with the CI that we were seeing before goes away (and if not, we may have something for him to look at). Thanks for your patience.

dpadhiar commented:
@juliev0 Addressed all review changes, I did see that intermittent failure a few times (I believe its when the ISBService is ready-ing) so I suppose we will still wait to merge on this.
